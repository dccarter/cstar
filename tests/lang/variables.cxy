// global variables
var X = 10;
const Y = ("Hello", "World");

func main(args: [string]) {
    var x;
    var y = 10;
    const k = y;
    x = k;
    // variable scope
    {
        var k = 10;
        var y = x;
    }
    // typed
    {
        var x: i32;
        var y: (bool, i32) = (true, 20);
    }
}