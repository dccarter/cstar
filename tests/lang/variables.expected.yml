Program:
decls:
- VarDecl:
  flags: (TopLevelDecl)
  names:
  - Identifier:
    value: X
  init:
    IntegerLit:
    value: 10
- VarDecl:
  flags: (Const|TopLevelDecl)
  names:
  - Identifier:
    value: Y
  init:
    TupleExpr:
    members:
    - StringLit:
      value: "Hello"
    - StringLit:
      value: "World"
- FuncDecl:
  flags: (TopLevelDecl)
  name: main
  params:
  - FuncParam:
    name: args
    index: 0
    paramType:
      ArrayType:
      flags: (TypeAst)
      element:
        StringType:
        flags: (TypeAst)
  body:
    BlockStmt:
    stmts:
    - VarDecl:
      names:
      - Identifier:
        value: x
    - VarDecl:
      names:
      - Identifier:
        value: y
      init:
        IntegerLit:
        value: 10
    - VarDecl:
      flags: (Const)
      names:
      - Identifier:
        value: k
      init:
        Path:
        elements:
        - PathElem:
          name: y
        isType: false
    - ExprStmt:
      expr:
        AssignExpr:
        op: 34
        lhs:
          Path:
          elements:
          - PathElem:
            name: x
          isType: false
        rhs:
          Path:
          elements:
          - PathElem:
            name: x
          isType: false
    - BlockStmt:
      stmts:
      - VarDecl:
        names:
        - Identifier:
          value: k
        init:
          IntegerLit:
          value: 10
      - VarDecl:
        names:
        - Identifier:
          value: y
        init:
          Path:
          elements:
          - PathElem:
            name: x
          isType: false
    - BlockStmt:
      stmts:
      - VarDecl:
        names:
        - Identifier:
          value: x
        varType:
          PrimitiveType:
          flags: (TypeAst)
          prtId: 9
      - VarDecl:
        names:
        - Identifier:
          value: y
        varType:
          TupleType:
          flags: (TypeAst)
          members:
          - PrimitiveType:
            flags: (TypeAst)
            prtId: 0
          - PrimitiveType:
            flags: (TypeAst)
            prtId: 9
        init:
          TupleExpr:
          members:
          - BoolLit:
            value: true
          - IntegerLit:
            value: 20
    - BlockStmt:
      stmts:
      - VarDecl:
        names:
        - Identifier:
          value: x
        init:
          IntegerLit:
          value: 20
      - VarDecl:
        names:
        - Identifier:
          value: y
        init:
          BoolLit:
          value: true
      - VarDecl:
        names:
        - Identifier:
          value: m
        varType:
          PrimitiveType:
          flags: (TypeAst)
          prtId: 9
      - VarDecl:
        names:
        - Identifier:
          value: n
        varType:
          PrimitiveType:
          flags: (TypeAst)
          prtId: 9
      - VarDecl:
        names:
        - Identifier:
          value: a
        init:
          IntegerLit:
          value: 20
      - VarDecl:
        names:
        - Identifier:
          value: b
        init:
          CharLit:
          value: ' '
      - VarDecl:
        names:
        - Identifier:
          value: d
        init:
          IntegerLit:
          value: 20
      - VarDecl:
        names:
        - Identifier:
          value: e
        init:
          CharLit:
          value: ' '
      - VarDecl:
        names:
        - Identifier:
          value: __gen_var_0
        init:
          CallExpr:
          callee:
            GroupExpr:
            expr:
              ClosureExpr:
              body:
                TupleExpr:
                members:
                - IntegerLit:
                  value: 20
                - IntegerLit:
                  value: 20
      - VarDecl:
        names:
        - Identifier:
          value: f
        init:
          MemberExpr:
          target:
            Path:
            elements:
            - PathElem:
              name: __gen_var_0
            isType: false
          member:
            IntegerLit:
            value: 0
      - VarDecl:
        names:
        - Identifier:
          value: g
        init:
          MemberExpr:
          target:
            Path:
            elements:
            - PathElem:
              name: __gen_var_0
            isType: false
          member:
            IntegerLit:
            value: 1
